File: py_CVRP/classes/GreedySolver.py
from classes import Route as rt
from classes.Solver import Solver
from classes.Solution import Solution


class GreedySolver(Solver):
    def __init__(self):
        super().__init__()

    def solve(self, problem):
        route_id = 0

        # As long as there are unvisited nodes
        while problem.any_location_unvisited():
            # Create a new route
            route_id += 1
            route = rt.Route(route_id, problem.vehicle_capacity)

            # Start at depot:
            current = problem.depot()
            route.add_waypoint(demand=current.demand,
                               distance=0,
                               waypoint_id=current.loc_id,
                               depot=True)
            problem.visit_depot()

            while True:
                # Then find the nearest unvisited customer who's demand can be fulfilled by the remaining cargo
                [found, nearest] = problem.nearest_unvisited(current, route.cargo_amount)

                # if there still is such a node:
                if found:
                    # add it to the route
                    route.add_waypoint(demand=nearest.demand,
                                       distance=problem.distance_between(
                                           current.loc_id, nearest.loc_id),
                                       waypoint_id=nearest.loc_id,
                                       depot=False)

                    # and mark it as visited
                    problem.visit_node(nearest.loc_id)

                    # then set start for next tour to current location
                    current = nearest

                # if there is no node left that can be visited on this route:
                else:
                    # return back to depot
                    nearest = problem.depot()
                    route.add_waypoint(demand=nearest.demand,
                                       distance=problem.distance_between(
                                           current.loc_id, nearest.loc_id),
                                       waypoint_id=nearest.loc_id,
                                       depot=True)
                    # add route to map
                    self.routes[route_id] = route
                    break

        for route in self.routes:
            print("Route #", self.routes[route].route_id, ":", self.routes[route].waypoints)
            self.total_cost += self.routes[route].total_cost

        print("Cost: ", self.total_cost)
        return Solution(self.routes, self.total_cost)

File: CVRP_cpp/source/greedy/GreedySolver.cpp
#include "GreedySolver.h"

GreedySolver::GreedySolver() = default;

GreedySolver::GreedySolver(LocationsContainer &locations, int capacity)
{
    _locations = locations;
    _vehicleCapacity = capacity;
}

void GreedySolver::Solve()
{
    std::vector <Route> routes{};
    int routeID = 0;

    while(_locations.AnyNodeUnvisited())
    {
        routeID += 1;
        // Start new Route at Depot
        Route route{_vehicleCapacity, routeID};
        LocationNode current = _locations.GetDepot();
        route.AddLocation(current);

        while (true)
        {
            //Then find the nearest unvisited node for which the _cargo is still enough
            const auto [found, nearest] = _locations.FindNearestUnvisited(current, route.Load());

            // If we did find a next node and the vehicle can satisfy its demand
            if(found && route.Load() - nearest.GetDemand() >= 0)
            {
                // add that node to the route
                route.AddLocation(nearest);

                // then mark the node as visited
                nearest.Visit();
            }
                // if we could not find a node that fulfills the conditions, we return to the depot
            else
            {
                current = _locations.GetDepot();
                route.AddLocation(current);

                // add the route to the solutions
                routes.push_back(route);

                // then start a new route
                break;
            }
        }
        int totalCost = 0;
        for(const auto& r : routes)
        {
            totalCost += r.GetTotalDistance();
            r.PrintRoute();
        }
        std::cout << "Cost " << totalCost << std::endl;

        bool visitedOnce = _locations.CheckVisitedOnce();
        if(visitedOnce)
        {
            std::cout << "In this Solution every Customer was visited exactly once." << std::endl;
        }
        else
        {
            std::cout << "There are customers missing from the Tour or some were visited multiple times." << std::endl;
        }

    }
}

